internal/util/metrics/metrics.go

```diff
@@ -0,0 +1,112 @@
+package metrics
+
+import (
+	"sync"
+
+	"github.com/prometheus/client_golang/prometheus"
+	"sigs.k8s.io/controller-runtime/pkg/metrics"
+)
+
+type CustomMetrics struct {
+	pipelinesGauge          *prometheus.GaugeVec
+	pipelineCounterMap      map[string]struct{}
+	isbsvcGauge             *prometheus.GaugeVec
+	isbsvcCounterMap        map[string]struct{}
+	numaflowControllerGauge *prometheus.GaugeVec
+}
+
+var (
+	pipelineLock sync.Mutex
+	isbsvcLock   sync.Mutex
+	// the pipelineMap contains the information of all running pipeline with key as "name_namespace"
+	pipelinesGauge = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Name: "numaflow_pipelines_running",
+			Help: "Number of Numaflow pipelines running",
+		},
+		[]string{},
+	)
+	isbsvcGauge = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Name: "numaflow_isb_service_running",
+			Help: "Number of Numaflow ISB Service running",
+		},
+		[]string{},
+	)
+	numaflowControllerGauge = prometheus.NewGaugeVec(
+		prometheus.GaugeOpts{
+			Name: "numaflow_controller_running",
+			Help: "Number of Numaflow controller running",
+		},
+		[]string{"name", "namespace", "version"},
+	)
+)
+
+// RegisterCustomMetrics registers the custom metrics to the existing global prometheus registry for pipelines, ISB service and numaflow controller
+func RegisterCustomMetrics() *CustomMetrics {
+	metrics.Registry.MustRegister(pipelinesGauge, isbsvcGauge, numaflowControllerGauge)
+
+	return &CustomMetrics{
+		pipelinesGauge:          pipelinesGauge,
+		pipelineCounterMap:      make(map[string]struct{}),
+		isbsvcGauge:             isbsvcGauge,
+		isbsvcCounterMap:        make(map[string]struct{}),
+		numaflowControllerGauge: numaflowControllerGauge,
+	}
+}
+
+// IncPipelineMetrics increments the pipeline counter
+func (m *CustomMetrics) IncPipelineMetrics(name, namespace string) {
+	pipelineLock.Lock()
+	defer pipelineLock.Unlock()
+	m.pipelineCounterMap[name+"_"+namespace] = struct{}{}
```

Comment:
So if the pipeline under the same name and namespace, it will not be added multiple times, right?

Reply:
right, it won't add any duplicate.